### ISSUE TITLE
Mattermost allows users to be removed from a DM (via the API) which breaks things

### ISSUE THREAD

**Issue by JonathonReinhart** [mattermost/mattermost-server#9579]
**TL;DR** My Mattermost system is broken because I tried to work-around a Mattermost bug.

It looks like Mattermost allows, via the API, the ability to "leave" a DM. If you do, you can never DM with that group again.

I'm running Mattermost as bundled with GitLab Omnibus.

## Story
I have a user ("W", `192.168.1.120` and later `192.168.1.58`) who used to use some Mattermost-IRC bridge. (I can get details if required).

Back in March of this year (whatever version of Mattermost is bundled with GitLab 10.3), W was in a chat with E and they found themselves unable to chat:

From W's perspective:
![chat](https://user-images.githubusercontent.com/1916566/46562515-5b915400-c8ca-11e8-8e2f-ff07d2ef4e44.png)

From E's perspective:

![image pasted at 2018-3-21 15-59](https://user-images.githubusercontent.com/1916566/46562576-ab701b00-c8ca-11e8-83e7-7f7df38dd7c1.png)

I saw these errors in the log for W:
```
[2018/03/21 15:58:47 EDT] [EROR] /api/v4/channels/b44zmm8aofykx8w98fydkdsnjo/members/me:Permissions code=403 rid=fnocu7okftda8fw9dq6yigseqa uid=ndfg5f8u1fn97mwhrerdfnj8tw ip=192.168.1.120 You do not have the appropriate permissions [details: userId=ndfg5f8u1fn97mwhrerdfnj8tw, permission=read_channel]
[2018/03/21 15:58:47 EDT] [EROR] /api/v4/channels/b44zmm8aofykx8w98fydkdsnjo/posts:Permissions code=403 rid=qg1jow9oq7yx5b98j4xwbgxo1w uid=ndfg5f8u1fn97mwhrerdfnj8tw ip=192.168.1.120 You do not have the appropriate permissions [details: userId=ndfg5f8u1fn97mwhrerdfnj8tw, permission=read_channel]
[2018/03/21 15:58:47 EDT] [EROR] /api/v4/channels/b44zmm8aofykx8w98fydkdsnjo/stats:Permissions code=403 rid=96eswawy47yz9na1gecx6jdoar uid=ndfg5f8u1fn97mwhrerdfnj8tw ip=192.168.1.120 You do not have the appropriate permissions [details: userId=ndfg5f8u1fn97mwhrerdfnj8tw, permission=read_channel]
[2018/03/21 15:58:49 EDT] [EROR] /api/v4/posts:Permissions code=403 rid=171nz8fhk3dc8ygyxoy3wkz7wa uid=ndfg5f8u1fn97mwhrerdfnj8tw ip=192.168.1.120 You do not have the appropriate permissions [details: userId=ndfg5f8u1fn97mwhrerdfnj8tw, permission=create_post]
[2018/03/21 15:58:53 EDT] [EROR] /api/v4/posts:Permissions code=403 rid=shyjxbo6ainx9q957o6uoq8qma uid=ndfg5f8u1fn97mwhrerdfnj8tw ip=192.168.1.120 You do not have the appropriate permissions [details: userId=ndfg5f8u1fn97mwhrerdfnj8tw, permission=create_post]
```

I had E help me troubleshoot this by setting their channel header to `SENTINEL`. Then from the `psql` console I did:
- Get the channel ID: `SELECT id FROM channels WHERE header LIKE '%SENTINEL%';`
- Get E and W's user IDs: `SELECT id,username FROM users WHERE username IN (`wxxxx`, `exxxx`);
- Look at the channel members: `SELECT * FROM channelmembers WHERE channelid='abc1234xyz';`
  - I confirmed that E was the only one in E/W's private DM.
- I "fixed" this by manually inserting W back into the `channelmembers`, copying all of E's column data (except `userid` of course).

This seemed to work okay. We wrote it off as a fluke, and he stopped using the Mattermost bridge.

------

Today, W, M, and I tried to have a 3-way DM.  (I'd already "closed" it, but M re-created it). (GitLab 11.2, Mattermost 5.1.1)

I saw in the history that W had "left" the channel back in February. (Again, you shouldn't be able to do this with a DM. :rotating_light:)

I tried the same "fix" as before:
   - Oops: `INSERT INTO channelmembers (channelid, userid, lastviewedat, msgcount, notifyprops, lastupdateat, schemeuser, schemeadmin) VALUES ('xxsntobdotb4tjhn7qfc4frx6e', 'ndfg5f8u1fn97mwhrerdfnj8tw', 1538772808650, 27, '"desktop":"default","email":"default","mark_unread":"all","push":"default"}', 1538772808650, 't', 'f');`
   - I accidentally inserted `notifyprops` without the leading `{`, and omitted `mentioncount` (which is nullable)
   - I corrected that mistake: `UPDATE channelmembers SET notifyprops='{"desktop":"default","email":"default","mark_unread":"all","push":"default"}', mentioncount=0 WHERE channelid='xxsntobdotb4tjhn7qfc4frx6e' AND userid='ndfg5f8u1fn97m    whrerdfnj8tw';`

## :bomb: 

Now, W has no Public Channels (other than Town Square), no Private Channels, and no Direct Messages in his web UI :warning:

![bad](https://user-images.githubusercontent.com/1916566/46563125-c09a7900-c8cd-11e8-9488-cb3275988774.png)

I see these errors in `/var/log/gitlab/mattermost/mattermost.log`:
```
{"level":"error","ts":1538776655.3192358,"caller":"app/web_conn.go:338","msg":"webhub.shouldSendEvent: SqlChannelStore.GetAllChannelMembersForUser: We couldn't get the channels, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 1: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776687.4998343,"caller":"app/web_conn.go:338","msg":"webhub.shouldSendEvent: SqlChannelStore.GetAllChannelMembersForUser: We couldn't get the channels, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 1: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776690.1470087,"caller":"web/handlers.go:142","msg":"We couldn't get the channel members","path":"/api/v4/users/me/teams/4y1gdkpd9jy39fzuym4iikgfxc/channels/members","request_id":"3kstnfbu3tdpjffupskdipiziy","ip_addr":"192.168.1.58","user_id":"ndfg5f8u1fn97mwhrerdfnj8tw","method":"GET","err_where":"SqlChannelStore.GetMembersForUser","http_code":500,"err_details":"teamId=4y1gdkpd9jy39fzuym4iikgfxc, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 2: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776724.924852,"caller":"app/web_conn.go:338","msg":"webhub.shouldSendEvent: SqlChannelStore.GetAllChannelMembersForUser: We couldn't get the channels, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 1: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776724.9297986,"caller":"app/web_conn.go:338","msg":"webhub.shouldSendEvent: SqlChannelStore.GetAllChannelMembersForUser: We couldn't get the channels, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 1: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776764.853685,"caller":"web/handlers.go:142","msg":"We couldn't get the channel members","path":"/api/v4/users/me/teams/4y1gdkpd9jy39fzuym4iikgfxc/channels/members","request_id":"qg6bmqq14pncjjpratdry1gbce","ip_addr":"192.168.1.58","user_id":"ndfg5f8u1fn97mwhrerdfnj8tw","method":"GET","err_where":"SqlChannelStore.GetMembersForUser","http_code":500,"err_details":"teamId=4y1gdkpd9jy39fzuym4iikgfxc, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 2: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776826.1814919,"caller":"app/web_conn.go:338","msg":"webhub.shouldSendEvent: SqlChannelStore.GetAllChannelMembersForUser: We couldn't get the channels, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 1: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776826.1863275,"caller":"app/web_conn.go:338","msg":"webhub.shouldSendEvent: SqlChannelStore.GetAllChannelMembersForUser: We couldn't get the channels, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 1: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776827.479852,"caller":"web/handlers.go:142","msg":"We couldn't get the channel members","path":"/api/v4/users/me/teams/4y1gdkpd9jy39fzuym4iikgfxc/channels/members","request_id":"ettqokh83fn3frc35bts9asxoa","ip_addr":"192.168.1.58","user_id":"ndfg5f8u1fn97mwhrerdfnj8tw","method":"GET","err_where":"SqlChannelStore.GetMembersForUser","http_code":500,"err_details":"teamId=4y1gdkpd9jy39fzuym4iikgfxc, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 2: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776878.2818816,"caller":"web/handlers.go:142","msg":"We couldn't get the channel members","path":"/api/v4/users/me/teams/4y1gdkpd9jy39fzuym4iikgfxc/channels/members","request_id":"6eni5aetstb37esinmdcopfeae","ip_addr":"192.168.1.58","user_id":"ndfg5f8u1fn97mwhrerdfnj8tw","method":"GET","err_where":"SqlChannelStore.GetMembersForUser","http_code":500,"err_details":"teamId=4y1gdkpd9jy39fzuym4iikgfxc, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 2: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776878.2878952,"caller":"web/handlers.go:142","msg":"You do not have the appropriate permissions","path":"/api/v4/channels/k8widm9iip8gud66wdgkfkjaba/posts","request_id":"w6dyyecmcfbwjxxrzdc8douere","ip_addr":"192.168.1.58","user_id":"ndfg5f8u1fn97mwhrerdfnj8tw","method":"GET","err_where":"Permissions","http_code":403,"err_details":"userId=ndfg5f8u1fn97mwhrerdfnj8tw, permission=read_channel"}
{"level":"error","ts":1538776892.7706215,"caller":"app/web_conn.go:338","msg":"webhub.shouldSendEvent: SqlChannelStore.GetAllChannelMembersForUser: We couldn't get the channels, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 1: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776892.7753289,"caller":"app/web_conn.go:338","msg":"webhub.shouldSendEvent: SqlChannelStore.GetAllChannelMembersForUser: We couldn't get the channels, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 1: unsupported Scan, storing driver.Value type <nil> into type *string"}
{"level":"error","ts":1538776897.2291765,"caller":"web/handlers.go:142","msg":"We couldn't get the channel members","path":"/api/v4/users/me/teams/4y1gdkpd9jy39fzuym4iikgfxc/channels/members","request_id":"hxefm7iua3rsbyxfga58bpw5ch","ip_addr":"192.168.1.58","user_id":"ndfg5f8u1fn97mwhrerdfnj8tw","method":"GET","err_where":"SqlChannelStore.GetMembersForUser","http_code":500,"err_details":"teamId=4y1gdkpd9jy39fzuym4iikgfxc, userId=ndfg5f8u1fn97mwhrerdfnj8tw, err=sql: Scan error on column index 2: unsupported Scan, storing driver.Value type <nil> into type *string"}
```

Here's the code for the SQL error:
https://github.com/mattermost/mattermost-server/blob/v5.1.1/app/web_conn.go#L338

So:
1. What is the bug that allows users to remove themselves from DMs, and not be allowed back in?
2. How can I fix my database for W?

**Comment by JonathonReinhart** [comment]
Here's the full log of my actions today, with user ID's changed for privacy, and a few huge `select *` queries removed: [full_log.txt](https://github.com/mattermost/mattermost-server/files/2452473/full_log.txt)

**Comment by amyblais** [comment]
Hi @JonathonReinhart, thank you for reaching out.

We fixed a bug for this in v5.3, so I recommend upgrading to that version - ticket here for reference: https://mattermost.atlassian.net/browse/MM-11593.

**Comment by JonathonReinhart** [comment]
Thanks @amyblais. I'm running Mattermost as part of GitLab so I'll have to wait for that update.

As far as the database goes, is there any way to recover this? It appears that there may be some other underlying DB issue that's I've inadvertently triggered.

**Comment by JonathonReinhart** [comment]
With the help of a SQL wizard colleague, we fixed it.

**The problem was that my `INSERT INTO channelmembers` did not include the `roles` column.**

When I first issued the `SELECT`, I saw that `roles` was "empty", so I omitted it from the `INSERT`.
```
# select * from channelmembers where channelid='xxsntobdotb4tjhn7qfc4frx6e';
         channelid          |           userid           | roles | lastviewedat  | msgcount | mentioncount |                                 notifyprops                                  | lastupdateat  | schemeuser | schemeadmin 
----------------------------+----------------------------+-------+---------------+----------+--------------+------------------------------------------------------------------------------+---------------+------------+-------------
 xxsntobdotb4tjhn7qfc4frx6e | ndfg5f8u1fn97mwhrerdfnj8tw |       | 1538773459750 |       33 |            0 | {"desktop":"default","email":"default","mark_unread":"all","push":"default"} | 1538773459750 | t          | f
 xxsntobdotb4tjhn7qfc4frx6e | c5x8xmjf8788umd3k3esgpeeae |       | 1538773524151 |       34 |            0 | {"desktop":"default","email":"default","mark_unread":"all","push":"default"} | 1538773524151 | t          | f
 xxsntobdotb4tjhn7qfc4frx6e | 4xc9tnz8kfr1pbin9peosfet9c |       | 1538773524151 |       34 |            0 | {"desktop":"default","email":"default","mark_unread":"all","push":"default"} | 1538773524151 | t          | f
(3 rows)
```

Recall the schema of the `channelmembers` table:
```
mattermost_production=# \d+ channelmembers
                               Table "public.channelmembers"
    Column    |          Type           | Modifiers | Storage  | Stats target | Description 
--------------+-------------------------+-----------+----------+--------------+-------------
 channelid    | character varying(26)   | not null  | extended |              | 
 userid       | character varying(26)   | not null  | extended |              | 
 roles        | character varying(64)   |           | extended |              | 
 lastviewedat | bigint                  |           | plain    |              | 
 msgcount     | bigint                  |           | plain    |              | 
 mentioncount | bigint                  |           | plain    |              | 
 notifyprops  | character varying(2000) |           | extended |              | 
 lastupdateat | bigint                  |           | plain    |              | 
 schemeuser   | boolean                 |           | plain    |              | 
 schemeadmin  | boolean                 |           | plain    |              | 
Indexes:
    "channelmembers_pkey" PRIMARY KEY, btree (channelid, userid)
    "idx_channelmembers_channel_id" btree (channelid)
    "idx_channelmembers_user_id" btree (userid)
```

**The problem is that `roles` should be marked `not null` because Go does not support scanning `null` values into string objects, hence the error:**

> `Scan error on column index 2: unsupported Scan, storing driver.Value type <nil> into type *string`

The fix was to update the row to set `roles=''`.

Furthermore, it was hard to detect this problem, because by default `psql` shows `null` values as empty strings.  I've since learned to use `\pset null '(null)'` to display null values as `(null)`.

**Mattermost should:**
- [ ] Ensure all `character varying` columns are marked `not null`, unless the code is explicitly written to handle null values:
   - Specifically `channelmembers.roles` in this case
- [ ] Set default values of empty string for these columns.
